File: tide_graph.php
# tide_graph.php
<?php
define('AMPLITUDE', 1.5);
define('MEAN_LEVEL', 0.5);
define('PERIOD', 12.42);

function tideLevel($hours) {
    $rad = ($hours / PERIOD) * 2 * M_PI;
    return MEAN_LEVEL + AMPLITUDE * sin($rad - 1.2);
}

function generateData() {
    $data = [];
    for ($minutes = 0; $minutes < 24*60; $minutes += 10) {
        $h = $minutes / 60.0;
        $lvl = tideLevel($h);
        $data[] = [$h, $lvl];
    }
    return $data;
}

function findExtrema($data) {
    $highs = $lows = [];
    for ($i = 1; $i < count($data)-1; $i++) {
        $prev = $data[$i-1];
        $curr = $data[$i];
        $next = $data[$i+1];
        if ($curr[1] > $prev[1] && $curr[1] > $next[1]) $highs[] = $curr;
        elseif ($curr[1] < $prev[1] && $curr[1] < $next[1]) $lows[] = $curr;
    }
    return [$highs, $lows];
}

function drawGraph($data) {
    $levels = array_column($data, 1);
    $min_lvl = min($levels) - 0.2;
    $max_lvl = max($levels) + 0.2;

    $height = 20;
    $scale = ($max_lvl - $min_lvl) / $height;

    $cols = 50;
    $step = floor(count($data) / $cols);
    $graphData = [];
    for ($i = 0; $i < $cols; $i++) {
        $idx = $i * $step;
        $h = $data[$idx][0];
        $lvl = $data[$idx][1];
        $row = round(($max_lvl - $lvl) / $scale);
        if ($row < 0) $row = 0;
        if ($row >= $height) $row = $height - 1;
        $graphData[] = ['h' => $h, 'lvl' => $lvl, 'row' => $row];
    }

    $lines = [];
    $lines[] = "🌊 Tide Graph (24‑hour forecast)";
    $lines[] = "";
    $lines[] = "Level (m)";

    for ($r = 0; $r < $height; $r++) {
        $lvl = $max_lvl - $r * $scale;
        $line = sprintf("%5.1f |", $lvl);
        for ($c = 0; $c < $cols; $c++) {
            $line .= $graphData[$c]['row'] == $r ? "*" : " ";
        }
        $lines[] = $line;
    }

    $lines[] = "      +" . str_repeat("-", $cols);

    $timeLine = "       ";
    for ($i = 0; $i <= 24; $i += 3) {
        $pos = (int)($i / 24 * $cols);
        $label = sprintf("%02d:00", $i);
        while (strlen($timeLine) < $pos) $timeLine .= " ";
        $timeLine .= $label;
    }
    $lines[] = $timeLine;
    $lines[] = "";

    list($highs, $lows) = findExtrema($data);
    if (!empty($highs)) {
        $str = "High tides: ";
        foreach (array_slice($highs, 0, 2) as $idx => $h) {
            if ($idx > 0) $str .= ", ";
            $hour = (int)$h[0];
            $min = round(($h[0] - $hour) * 60);
            $str .= sprintf("%02d:%02d (%.1fm)", $hour, $min, $h[1]);
        }
        $lines[] = $str;
    }
    if (!empty($lows)) {
        $str = "Low tides:  ";
        foreach (array_slice($lows, 0, 2) as $idx => $l) {
            if ($idx > 0) $str .= ", ";
            $hour = (int)$l[0];
            $min = round(($l[0] - $hour) * 60);
            $str .= sprintf("%02d:%02d (%.1fm)", $hour, $min, $l[1]);
        }
        $lines[] = $str;
    }

    return implode("\n", $lines);
}

echo drawGraph(generateData()) . "\n";
?>

File: TideGraph.cs
// TideGraph.cs
using System;
using System.Collections.Generic;
using System.Linq;

class TideGraph
{
    const double AMPLITUDE = 1.5;
    const double MEAN_LEVEL = 0.5;
    const double PERIOD = 12.42;

    static double TideLevel(double hours)
    {
        double rad = (hours / PERIOD) * 2 * Math.PI;
        return MEAN_LEVEL + AMPLITUDE * Math.Sin(rad - 1.2);
    }

    static List<(double hour, double level)> GenerateData()
    {
        var data = new List<(double, double)>();
        for (int minutes = 0; minutes < 24*60; minutes += 10)
        {
            double h = minutes / 60.0;
            data.Add((h, TideLevel(h)));
        }
        return data;
    }

    static string DrawGraph(List<(double hour, double level)> data)
    {
        var levels = data.Select(d => d.level).ToList();
        double minLvl = levels.Min() - 0.2;
        double maxLvl = levels.Max() + 0.2;

        int height = 20;
        double scale = (maxLvl - minLvl) / height;

        int cols = 50;
        int step = data.Count / cols;
        int[] rows = new int[cols];
        double[] times = new double[cols];
        double[] lvls = new double[cols];
        for (int i = 0; i < cols; i++)
        {
            int idx = i * step;
            times[i] = data[idx].hour;
            lvls[i] = data[idx].level;
            int row = (int)Math.Round((maxLvl - lvls[i]) / scale);
            if (row < 0) row = 0;
            if (row >= height) row = height - 1;
            rows[i] = row;
        }

        var lines = new List<string>();
        lines.Add("🌊 Tide Graph (24‑hour forecast)");
        lines.Add("");
        lines.Add("Level (m)");

        for (int r = 0; r < height; r++)
        {
            double lvl = maxLvl - r * scale;
            string line = $"{lvl,5:F1} |";
            for (int c = 0; c < cols; c++)
                line += rows[c] == r ? "*" : " ";
            lines.Add(line);
        }

        lines.Add("      +" + new string('-', cols));

        string timeLine = "       ";
        for (int i = 0; i <= 24; i += 3)
        {
            int pos = (int)((i / 24.0) * cols);
            string label = $"{i:D2}:00";
            while (timeLine.Length < pos) timeLine += " ";
            timeLine += label;
        }
        lines.Add(timeLine);
        lines.Add("");

        // Extrema
        var highs = new List<(double hour, double level)>();
        var lows = new List<(double hour, double level)>();
        for (int i = 1; i < data.Count - 1; i++)
        {
            if (data[i].level > data[i-1].level && data[i].level > data[i+1].level)
                highs.Add(data[i]);
            else if (data[i].level < data[i-1].level && data[i].level < data[i+1].level)
                lows.Add(data[i]);
        }

        if (highs.Any())
        {
            string str = "High tides: ";
            for (int i = 0; i < Math.Min(2, highs.Count); i++)
            {
                if (i > 0) str += ", ";
                int hour = (int)highs[i].hour;
                int min = (int)Math.Round((highs[i].hour - hour) * 60);
                str += $"{hour:D2}:{min:D2} ({highs[i].level:F1}m)";
            }
            lines.Add(str);
        }
        if (lows.Any())
        {
            string str = "Low tides:  ";
            for (int i = 0; i < Math.Min(2, lows.Count); i++)
            {
                if (i > 0) str += ", ";
                int hour = (int)lows[i].hour;
                int min = (int)Math.Round((lows[i].hour - hour) * 60);
                str += $"{hour:D2}:{min:D2} ({lows[i].level:F1}m)";
            }
            lines.Add(str);
        }

        return string.Join("\n", lines);
    }

    static void Main()
    {
        var data = GenerateData();
        Console.Write(DrawGraph(data));
    }
}
